lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/address/ContainerBasedLocalAddressProvider.java

```diff
@@ -0,0 +1,110 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.reef.wake.remote.address;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.reef.tang.Configuration;
+import org.apache.reef.tang.Tang;
+import org.apache.reef.tang.exceptions.InjectionException;
+import org.apache.reef.wake.remote.RemoteConfiguration;
+
+import javax.inject.Inject;
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * A LocalAddressProvider that reads the contents of the file at HOST_IP_ADDR_PATH and uses it to be the ip address.
+ */
+public final class ContainerBasedLocalAddressProvider implements LocalAddressProvider {
+
+  public static final String HOST_IP_ADDR_PATH_ENV = "HOST_IP_ADDR_PATH";
+  private static final Logger LOG = Logger.getLogger(ContainerBasedLocalAddressProvider.class.getName());
+
+  private String cachedLocalAddress = null;
+
+  /**
+   * The constructor is for Tang only.
+   */
+  @Inject
+  private ContainerBasedLocalAddressProvider() {
+    LOG.log(Level.FINE, "Instantiating ContainerBasedLocalAddressProvider");
+  }
+
+  @Override
+  public synchronized String getLocalAddress() {
+    if (cachedLocalAddress != null) {
+      return cachedLocalAddress;
+    }
+
+    String ipAddressPath = System.getenv(HOST_IP_ADDR_PATH_ENV);
+    LOG.log(Level.FINE, "IpAddressPath is {0}", ipAddressPath);
+    if (StringUtils.isEmpty(ipAddressPath)) {
+      final String message = String.format("Environment variable must be set for %s", HOST_IP_ADDR_PATH_ENV);
+      LOG.log(Level.SEVERE, message);
+      throw new RuntimeException(message);
+    }
+
+    File ipAddressFile = new File(ipAddressPath);
+    if (!ipAddressFile.exists() || !ipAddressFile.isFile()) {
+      final String message = String.format("%s points to invalid path: %s", HOST_IP_ADDR_PATH_ENV, ipAddressPath);
+      LOG.log(Level.SEVERE, message);
+      throw new RuntimeException(message);
+    }
+
+    try {
+      cachedLocalAddress = readFile(ipAddressPath, StandardCharsets.UTF_8);
+      return cachedLocalAddress;
+    } catch (IOException e) {
+      String message = String.format("Exception when attempting to read file %s", ipAddressPath);
+      LOG.log(Level.SEVERE, message, e);
+      throw new RuntimeException(message, e);
+    }
+  }
+
+  @Override
+  public Configuration getConfiguration() {
+    try {
+      LocalAddressProvider defaultLocalAddressProvider =
+          Tang.Factory.getTang().newInjector().getInstance(LocalAddressProvider.class);
+      return Tang.Factory.getTang().newConfigurationBuilder()
+          .bind(LocalAddressProvider.class, ContainerBasedLocalAddressProvider.class)
+          .bindNamedParameter(RemoteConfiguration.HostAddress.class, defaultLocalAddressProvider.getLocalAddress())
```

Comment:
>RemoteConfiguration.HostAddress.class [](start = 30, length = 37)

Is this needed?

Reply:
Removed.

---
In reply to: [199298882](https://github.com/sharathmalladi/reef/pull/6#discussion_r199298882) [](ancestors = 199298882)